### PR TITLE
SASS-2452 did you pay non-uk tax surcharge page

### PR DESCRIPTION
**Pull request by walesoares**
### Description
Please include a brief summary of the change you've made and the issue that it fixes.

Add a link to the relevant story in Jira
[SASS-2452](https://jira.tools.tax.service.gov.uk/browse/SASS-2452)

##### Before Reviewing
- [x]  Have you pulled the branch down?
- [x]  Have you assigned yourself to the PR?
- [x]  Have you moved the task to “in review” on JIRA?
- [x]  Have you checked to ensure all dependencies are up to date?
- [ ]  Have you checked to ensure its been rebased against the current version of main?

##### Whilst Reviewing
- [x]  Have you run the tests?
- [ ]  Have you run the journey tests?
- [x]  Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?

##### After Reviewing
- [ ]  Have you checked for merge conflicts or any changes in the current main that may affect the current pull request? i.e. does it need another rebase?
- [ ]  Have you checked to make sure there are no builds in the pipeline before you merge?
- [ ]  Have you moved the task to “in pipeline” on Jira?

##### Before creating PR
- [x]  Have you run the tests?
- [ ]  Have you run the journey tests? (where applicable)
- [x]  Have you addressed warnings where appropriate?
- [x]  Have you rebased against the current version of main?
- [x]  Have you checked code coverage isn’t lower than previously?

##### After PRs been raised
- [x]  Have you checked the PR Builder passes?

### CODE REVIEW

**Comment by erasmos-hmrc** [comment]
I did some QA as both the individual and agent, and all works according to the spec.

**Comment by erasmos-hmrc** [comment]
@abiodxn, could you rebase please?

File: app/forms/RadioButtonAmountForm.scala [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ */
+
+package forms
+
+import forms.validation.mappings.MappingUtil.optionCurrency
+import play.api.data.Forms.{of, tuple}
+import play.api.data.format.Formatter
+import play.api.data.{FieldMapping, Form, FormError}
+
+object RadioButtonAmountForm {
```
Comment (erasmos-hmrc) [inline app/forms/RadioButtonAmountForm.scala:24]: Should this have its own unit tests?

File: app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+
+  def submit(taxYear: Int): Action[AnyContent] = authAction.async { implicit request => {
+    pensionSessionService.getPensionsSessionDataResult(taxYear, request.user) {
+      case Some(data) =>
+        if (data.pensions.unauthorisedPayments.surchargeAmount.isDefined) {
+          form.bindFromRequest.fold(
+            formWithErrors => Future.successful(BadRequest(didYouPayNonUkTaxSurchargeView(formWithErrors, taxYear))),
+            amounts => {
```
Comment (erasmos-hmrc) [inline app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala:87]: You could use a case here, with both parts of the tuple.

```
case (yesSelected, amountOpt) => {
    val pensionsCYAModel: PensionsCYAModel = data.pensions
    val viewModel: UnauthorisedPaymentsViewModel = pensionsCYAModel.unauthorisedPayments
    val updatedCyaModel: PensionsCYAModel = {
      pensionsCYAModel.copy(
        unauthorisedPayments = viewModel.copy(
          surchargeTaxAmountQuestion = Some(yesSelected),
          surchargeTaxAmount = if (yesSelected) amountOpt else Some(0)
        )
      )
    }
    pensionSessionService.createOrUpdateSessionData(request.user,
      updatedCyaModel, taxYear, data.isPriorSubmission)(errorHandler.internalServerError()) {
      //TODO: next page (were any of the unauthorised payments from a uk pension scheme page)
      Redirect(DidYouPayNonUkTaxController.show(taxYear))
    }
  }
```
That way, we don't have to reference the tuple parts by index.

File: app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    emptyFieldKey = s"unauthorisedPayments.didYouPayNonUkTax.error.Amount.noEntry",
+    wrongFormatKey = s"unauthorisedPayments.didYouPayNonUkTax.error.Amount.incorrectFormat",
+    exceedsMaxAmountKey = s"unauthorisedPayments.didYouPayNonUkTax.error.Amount.overMaximum"
+  )
+
+
+  def show(taxYear: Int): Action[AnyContent] = authAction.async { implicit request => {
+    pensionSessionService.getPensionsSessionDataResult(taxYear, request.user) {
+      case Some(data) =>
+        if (data.pensions.unauthorisedPayments.surchargeAmount.isDefined) {
```
Comment (erasmos-hmrc) [inline app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala:57]: You could combine this line with the one above, as a guard condition, as you do on line 59.

**Review by erasmos-hmrc — COMMENTED** [review]
It works as it should. Only added suggestions.

**Review by erasmos-hmrc — DISMISSED** [review]

File: app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala (hunk 2 above)
Comment (erasmos-hmrc) [reply app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala:87]: Sorry - I should have compiled that. Ignore for now.

Comment (abiodxn) [reply app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala:87]: Not sure why but the type isn't resolving on my machine it says `yesSelected` is of type `Any`

File: app/forms/RadioButtonAmountForm.scala (hunk 1 above)
Comment (abiodxn) [reply app/forms/RadioButtonAmountForm.scala:24]: I was thinking to continue as we're going (as we dont' have unit tests for forms as they're tested in the IT test) and then apply the unit tests once we pick up the task to move the integration tests to unit test

File: app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala (hunk 2 above)
Comment (erasmos-hmrc) [reply app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala:87]: This compiles, and all tests pass:

```
            amounts => {
              val pensionsCYAModel: PensionsCYAModel = data.pensions
              val viewModel: UnauthorisedPaymentsViewModel = pensionsCYAModel.unauthorisedPayments
              val updatedCyaModel: PensionsCYAModel = amounts match {
                case (yesSelected, amountOpt) =>
                  pensionsCYAModel.copy(
                    unauthorisedPayments = viewModel.copy(
                      surchargeTaxAmountQuestion = Some(yesSelected),
                      surchargeTaxAmount = if (yesSelected) amountOpt else Some(0)
                    )
                  )
              }
              pensionSessionService.createOrUpdateSessionData(request.user,
                updatedCyaModel, taxYear, data.isPriorSubmission)(errorHandler.internalServerError()) {
                //TODO: next page (were any of the unauthorised payments from a uk pension scheme page)
                Redirect(DidYouPayNonUkTaxController.show(taxYear))
              }
            }
```

File: app/forms/RadioButtonAmountForm.scala (hunk 1 above)
Comment (erasmos-hmrc) [reply app/forms/RadioButtonAmountForm.scala:24]: OK.

File: app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala (hunk 2 above)
Comment (erasmos-hmrc) [reply app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala:87]: It doesn't. The second one is correct.

Comment (erasmos-hmrc) [reply app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala:85]: Confirmed.

**Review by erasmos-hmrc — APPROVED** [review]

**Pull request merged**